Model.py

```diff
@@ -0,0 +1,129 @@
+import os
+import cv2
+import torch
+import numpy as np
+from facenet_pytorch import InceptionResnetV1, MTCNN
+from types import MethodType
+
+### helper function
+def encode(img):
+ res = resnet(torch.Tensor(img))
+ return res
+
+# Custom face detection method
+def detect_box(self, img, save_path=None):
+  batch_boxes, batch_probs, batch_points = self.detect(img, landmarks=True)
+ 
+  # Select Faces (Skipping small ones)
+  if not self.keep_all:
+    small_face_indices = [i for i, box in enumerate(batch_boxes) if box[2] - box[0] < self.min_face_size]
+    batch_boxes = np.delete(batch_boxes, small_face_indices, axis=0)
+    batch_probs = np.delete(batch_probs, small_face_indices, axis=0)
+    if len(batch_points) > 0:
+      batch_points = np.delete(batch_points, small_face_indices, axis=0)
+
+# Extract faces
+  faces = self.extract(img, batch_boxes, save_path)
+  return batch_boxes, faces
+
+### load model
+resnet = InceptionResnetV1(pretrained='vggface2').eval()
+mtcnn = MTCNN(margin=44,image_size=182, keep_all=True, thresholds=[0.4, 0.5, 0.5], min_face_size=60)
+mtcnn.detect_box = MethodType(detect_box, mtcnn)
+
+# Initialize a list to store distances for dynamic thresholding
+known_distances = []
+
+### load known faces
+saved_pictures = "./C3_Members/"    ### path to folder containing pictures of known faces
+all_people_faces = {}
+
+# Define detection frequency
+detection_counter = 0
+detection_frequency = 3  # Adjust this value for your desired balance between accuracy and speed
+
+# Iterate through folders containing individual's pictures
+for person_folder in os.listdir(saved_pictures):
+ person_path = os.path.join(saved_pictures, person_folder)
+ for file in os.listdir(person_path):
+  if file.endswith(".jpeg"):
+   img = cv2.imread(os.path.join(person_path, file))
+   cropped = mtcnn(img)
+   if cropped is not None:
+    all_people_faces[person_folder] = encode(cropped)[0, :]
+    known_distances.append(encode(cropped)[0, :])
+
+# Calculate a dynamic threshold based on the distances in the known dataset
+known_distances = torch.stack(known_distances)
+dynamic_threshold = np.percentile([dist.norm().item() for dist in known_distances], 90) # Adjust percentile as needed
+     
+     
+# Initialize lists for storing predicted and true labels
+predicted_labels = []
+true_labels = []
+confidence_scores = [] # List to store confidence scores
+
+# Initialize a counter for correct predictions
+correct_predictions = 0
+total_predictions = 0
+
+### initialize webcam and capture loop
```

Comment:
Comment out all the lines from line 70 - 129(end) so that they can be used eventually for checking purposes using webcam.